### ISSUE TITLE
CFE generates invalid expressions for constructor tear-offs called in const context

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#46925]
On the latest https://github.com/dart-lang/co19/blob/98d359a8ff8e642679d695c2c5aecb0b0d7f7f70/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:

```
out/ReleaseX64/dart --enable-experiment=constructor-tearoffs tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart
Unhandled exception:
'file://.../tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart': error: function invocation
#0      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:283:19)
#1      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

As of today this version of the test hasn't been rolled into Dart SDK yet, so you might need to get it from the co19 repository.

Kernel file:
```
  static method main() → dynamic {
    const nam::MyClass<core::String> c1 = invalid-expression "function invocation";
    const nam::MyClass<core::String> c2 = invalid-expression "function invocation";
    Exp::Expect::identical(invalid-expression "function invocation", invalid-expression "function invocation");
    const nam::MyClass<core::int> c3 = invalid-expression "function invocation";
    const nam::MyClass<core::int> c4 = invalid-expression "function invocation";
    Exp::Expect::identical(invalid-expression "function invocation", invalid-expression "function invocation");
    Exp::Expect::notEquals(invalid-expression "function invocation", invalid-expression "function invocation");
    const nam::MyClass<core::int> c5 = invalid-expression "function invocation";
    Exp::Expect::notEquals(invalid-expression "function invocation", invalid-expression "function invocation");
    const nam::MyClass<core::String> c6 = invalid-expression "function invocation";
    Exp::Expect::notEquals(invalid-expression "function invocation", invalid-expression "function invocation");
  }
```

There are a couple of problems:

1) CFE shouldn't silently generate InvalidExpression nodes without reporting a compile-time error. There were no errors from the front-end, this exception was thrown by the VM after loading a kernel file.

2) Do we allow calling constructor tear-offs in the constant context? If we do, then CFE should support that. If we don't, then the test is incorrect and should be fixed.

/cc @johnniwinther @stefantsov @eernstg

**Comment by johnniwinther** [comment]
At bleeding edge I get
```
tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:28:14: Error: Not a constant expression.
  const c1 = v1(3, 14);
             ^^
tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:29:14: Error: Not a constant expression.
  const c2 = v1(3, 14);
             ^^
tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:32:14: Error: Not a constant expression.
  const c3 = v2();
             ^^
tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:33:14: Error: Not a constant expression.
  const c4 = v2();
             ^^
tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:38:14: Error: Not a constant expression.
  const c5 = v3(3, 14);
             ^^
tests/co19/src/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart:41:14: Error: Not a constant expression.
  const c6 = v4();
             ^^
```

**Comment by johnniwinther** [comment]
Regarding 1) : I agree that we should never generate an `InvalidExpression` without having reported a compile-time error. https://dart-review.googlesource.com/c/sdk/+/207001 has added an assertion for this.

**Comment by alexmarkov** [comment]
@johnniwinther Did you try the latest version of the test from co19 repository, namely https://github.com/dart-lang/co19/blob/98d359a8ff8e642679d695c2c5aecb0b0d7f7f70/LanguageFeatures/Constructor-tear-offs/named_constructor_A08_t02.dart ?

**Comment by johnniwinther** [comment]
No, I didn't. I can repro the error with `dart`, but using `_fasta/compile.dart`, I actually get the compile-time errors.

I'll look further into it.

**Comment by johnniwinther** [comment]
No, wrong command. `_fasta/compile.dart` doesn't provide a compile-time error, but it _does_ fail with the assertion.

**Comment by eernstg** [comment]
> Do we allow calling constructor tear-offs in the constant context

No, invocation of a constructor tearoff is never a constant expression (it denotes an invocation of a given function object, and that is in general going to perform arbitrary computations; a constructor can run arbitrary code, so that's fully justified).

We could make an exception for a constant expression that is a constructor tearoff _of a constant constructor_ (if it is generically instantiated then the type arguments are constant, because otherwise it wouldn't be a constant expression). However, I don't think that it would be a very useful feature, because we certainly won't allow arbitrary functions to be invoked during constant expression evaluation, and if we know that we are invoking a tearoff of a constant constructor then we could just as well avoid tearing it off in the first place, and call it directly.

The request for a change to the test was made in https://github.com/dart-lang/co19/issues/1151, and that issue is closed, so we're just waiting for a co19 roll.

**Comment by alexmarkov** [comment]
> The request for a change to the test was made in dart-lang/co19#1151, and that issue is closed, so we're just waiting for a co19 roll.

Note that this issue is about the new version of the test in the co19 repository, after dart-lang/co19#1151 was closed. That version is not rolled yet, but I started looking at the most recent version of co19 tests from co19 repository as rolls take some time. So maybe we should reopen the co19 issue or file another co19 issue about correcting the test.
